columnflow/production/pdf.py

```diff
@@ -0,0 +1,59 @@
+# coding: utf-8
+
+"""
+Column production methods related to the PDF weights.
+"""
+
+
+from columnflow.production import Producer, producer
+from columnflow.util import maybe_import
+from columnflow.columnar_util import set_ak_column
+
+ak = maybe_import("awkward")
+
+
+@producer(
+    uses={"LHEPdfWeight"},
+    produces={
+        "pdf_weight", "pdf_weight_up", "pdf_weight_down",
+    },
+)
+def pdf_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Producer that determines the pdf up and down variations on an event-by-event basis.
+    Can only be called with MC datasets.
+
+    Resources:
+
+       - https://arxiv.org/pdf/1510.03865.pdf
+    """
+
+    # stop here for data
+    if self.dataset_inst.is_data:
+        raise ValueError("attempt to determine pdf variations in data")
+
+    N_pdfweights = ak.num(events.LHEPdfWeight, axis=1)
+    if ak.any(N_pdfweights != 103) or ak.any(N_pdfweights != 101):
+        raise Exception(
+            f"Number of LHEPdfWeights ({N_pdfweights}) is not as expected (103 or 101) "
+            f"in dataset {self.dataset_inst.name}",
+        )
+
+    # first LHEPdfWeight value: nominal weight
+    pdf_weight_nominal = events.LHEPdfWeight[:, 0]
+    if ak.any(pdf_weight_nominal != 1):
+        print("The first entry of the LHEPdfWeight is not 1 but will be used as the nominal weight.")
+
+    # the following 100 LHEPdfWeight values: pdf variations
+    pdfweights = events.LHEPdfWeight[:, 1:101] / pdf_weight_nominal
+    pdfweights = ak.sort(pdfweights)
+
+    # PDF uncertainty as 68% CL
+    var = (pdfweights[:, 83] - pdfweights[:, 15]) / 2
+
+    # NOTE: use mean value as nominal pdf weight? or remove the necessity of adding this nominal weight?
+    events = set_ak_column(events, "pdf_weight", pdf_weight_nominal)
+    events = set_ak_column(events, "pdf_weight_down", pdf_weight_nominal + var)
+    events = set_ak_column(events, "pdf_weight_up", pdf_weight_nominal - var)
```

Comment:
Mh, if I see correctly, then `var` (or `std` with the above comment), should be scaled first by `pdf_weight_nominal` again. As it is right now, `var` is the standard deviation of the weights that were previously scaled to nominal.

This means that if nominal is != 1, you are adding / subtracting a normalized value (sth around 1) to / from a value that can be (e.g.) 0.0001.

I think the columns in lines 56/57 should be `pdf_weight_nominal * (1 +- var)`.

Reply:
you're right, or I just skip dividing by the nominal weight to begin with

Reply:
But then the "weight" would not be usable anymore as a multiplicative weight out of the box, right?

Reply:
yes, this is the case when all weights are normalised to one entry w_0 we know of, W_i = w_i / w_0. However, in some samples, the "w_0" is actually not the first entry in the list of pdf weights. So, assuming all weights are normalized to some w_N (where we don't know which weight this is or even if N is the same for all events), we cannot identify which weight the nominal weight is. So in my case, I then simply redefine the weight w_0 as the nominal weight by multiplying with W_0 = w_0 / w_N. When N=0, this should not change anything and when N!=0 we now use a different weight as the nominal one.